src/reader.rs

```diff
@@ -2,89 +2,126 @@ use bincode::Options;
 use byteorder::{ByteOrder, NetworkEndian};
 use bytes::buf::Buf;
 use bytes::BytesMut;
-use futures_core::{ready, Stream};
+use futures_core::ready;
 use serde::Deserialize;
 use std::io;
 use std::marker::PhantomData;
 use std::pin::Pin;
 use std::task::{Context, Poll};
-use tokio::io::{AsyncRead, ReadBuf};
-
-/// A wrapper around an asynchronous reader that produces an asynchronous stream of
-/// bincode-decoded values.
-///
-/// To use, provide a reader that implements [`AsyncRead`], and then use [`Stream`] to access the
-/// deserialized values.
-///
-/// Note that the sender *must* prefix each serialized item with its size as reported by
-/// `bincode::serialized_size` encoded as a four-byte network-endian encoded. See also
-/// [`serialize_into`], which does this for you.
-#[derive(Debug)]
-pub struct AsyncBincodeReader<R, T> {
-    reader: R,
-    pub(crate) buffer: BytesMut,
-    into: PhantomData<T>,
-}
 
-impl<R, T> Unpin for AsyncBincodeReader<R, T> where R: Unpin {}
+macro_rules! make_reader {
```

Comment:
The more I look at this macro, the more I feel like it _should_ be doable without a macro. I tried mocking something up on the playground and it _seems_ to work: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2d284f511e2fa13170515ed6bd6f373d

What do you think of that approach instead? Is it something you've tried and that turns out not to work (due to GATs or something along those lines)?

Reply:
@jonhoo It'd be fairly straightforward to do that for the reader code; I did go down a similar route at first. It seems much harder to get that working for the writer and stream code; even though the traits are similar for the two frameworks, a fair bit of similar code would still get duplicated twice.

I think it's *absolutely* possible, but I don't think it's going to be elegant.

I also don't think this would be as maintainable (though as the maintainer that's your call).

And, apart from that, I'm also hoping the half-life of this division doesn't last much longer; I hope version 0.8 (or 1.0) of this crate uses traits from std, and near-zero async-framework-specific code.

Reply:
Fair enough, I trust your assessment here :+1: 
I was thinking doing it with another generic argument might make it easier for _consumers_ of this crate to also be framework agnostic, but maybe that's not true either.

Let's land this as-is, and then we can always iterate further down the line should it be necessary.

And yes, can't wait for shared traits!

Reply:
> I was thinking doing it with another generic argument might make it easier for _consumers_ of this crate to also be framework agnostic, but maybe that's not true either.

Ah, I see! Yeah, I think that's theoretically possible, but it'd be a much larger engineering challenge to make that portability layer useful for other code, and I think that's a separate project.

(Also, a couple of portability layers already exist, and people tend to not want to use them because they're an extra layer/crate.)